modules/usrloc/dlist.c

```diff
@@ -254,13 +255,36 @@ static int get_domain_db_ucontacts(udomain_t *d, void *buf, int *len,
 				continue;
 			}
 
+			/* We cannot add items that point to p or p1 here,
+			 * unless they point into the buffer we're keeping.
+			 * So: we write to buf, but keep a pointer to the
+			 * start, so we can revert this record if needed. */
+			record_start = buf;
+
+			/* write received/contact */
+			memcpy(buf, &p_len, sizeof p_len);
+			buf += sizeof p_len;
+			memcpy(buf, p, p_len);
+			p = buf; /* point to to-be-kept copy of p */
+			buf += p_len;
+
+			/* write path */
+			memcpy(buf, &p1_len, sizeof p1_len);
+			buf += sizeof p1_len;
+			memcpy(buf, p1, (unsigned)p1_len);
+			p1 = buf; /* point to to-be-kept copy of p1 */
+			buf += p1_len;
+
 			/* determine and parse the URI of this contact's next hop */
 			if (p1_len > 0) {
 				/* send to first URI in path */
+				str uri, host;
```

Comment:
While I'm a big fan of block-scope variable declarations, what do you think is the best approach given that using them opens the doors to problems such as the one in 3ccea1a32?  Talk to Nick to drop those types of builds altogether or take it upon ourselves to refrain from assuming a C99+ standard for a couple more years?

Reply:
This is already legal in old C.

The only problem in 3ccea1a is `for (int i` that's standard C++ and C99+, but not ansi/C89.

See this example:
```c
int func() {
    int x;

    {
#ifndef OLD_C
        for (int i = 0; i < 10; ++i) {
#else
        int i;
        for (i = 0; i < 10; ++i) {
#endif
            float y;
            y += i;
            x = y;
        }
    }

    return x;
}
```
```console
$ gcc -Wall -ansi -c example.c
example.c: In function ‘func’:
example.c:6:9: error: ‘for’ loop initial declarations are only allowed in C99 or C11 mode
         for (int i = 0; i < 10; ++i) {
         ^~~
example.c:6:9: note: use option -std=c99, -std=gnu99, -std=c11 or -std=gnu11 to compile your code

$ gcc -DOLD_C -Wall -ansi -c example.c
(success)
```

Reply:
Brilliant!  Thank you! <3